runtime/backtrace.c

```diff
@@ -156,9 +158,15 @@ CAMLprim value caml_get_exception_raw_backtrace(value unit)
     memcpy(saved_backtrace_buffer, Caml_state->backtrace_buffer,
            saved_backtrace_pos * sizeof(backtrace_slot));
 
-    res = caml_alloc(saved_backtrace_pos, 0);
-    for (i = 0; i < saved_backtrace_pos; i++) {
-      Field(res, i) = Val_backtrace_slot(saved_backtrace_buffer[i]);
+    res = caml_alloc_shr_no_track_noexc(saved_backtrace_pos, 0);
+
+    if (res == 0) {
+      // out of memory
+      res = caml_alloc(0,0);
+    } else {
+      for (i = 0; i < saved_backtrace_pos; i++) {
+        Field(res, i) = Val_backtrace_slot(saved_backtrace_buffer[i]);
+      }
```

Comment:
This could use a comment explaining why it's ok to use Field on the result of caml_alloc_shr, given that the manual says specifically not to do this (https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html, rule 6), similar I think to some comments in backtrace_{byt,nat}.c.

(also: I don't know if it's intended, but I'll point out that this change makes this function always allocate on the major heap)

Reply:
> this change makes this function always allocate on the major heap

I'm afraid this is not acceptable.  A lot of effort went recently in making backtrace operations more efficient, and allocation in the minor heap when possible was part of this effort.

Reply:
I just added such a comment. Notice that this is unrelated to the patch, since it was already possible that the block was in the major heap before. Incidentally, the comment in backtrace_{byt,nat}.c was wrong because if it had been calling caml_modify, this would have been on uninitialised memory in one call site.

Thanks for pointing out that it never allocates on the minor heap, I completely forgot to mention it in the description and that I wanted to discuss it. Yes, this is by choice (to reuse an existing function). I briefly contemplated writing a variant of caml_alloc that returns 0 (or a value that satifies Is_exception_result to be consistent with the rest of the FFI) instead of raising. It could be of general use to the FFI users now that there only remains one situation where caml_alloc can directly raise into OCaml (out of memory). I would be willing to do that as part of a patch to provide a FFI interface safe for C-style resource management (where the user checks for exceptional result and re-raises after doing some clean-up). However in that use-case I judged it was fine to allocate on the major heap. I am only worried about repeated get_raw_backtrace/raise_with_backtrace that occurs with nested Fun.protect (for instance),  but this has to be addressed separately (one idea is to cache the pointer to the last allocated backtrace inside Caml_state).

Reply:
@xavierleroy just saw your comment. OK, we'll do it differently. Would any of the two suggestions (make the FFI generally non-raising, cache the allocated backtrace) provide an acceptable solution?

Reply:
Heh, apologies, I missed the `_shr` part of the allocator logic. I agree this change is problematic; we discussed it already in https://github.com/ocaml/ocaml/pull/8670#issuecomment-492568414 and following.

Reply:
Thanks for reminding me of this discussion. I thought further about what should be done in the unwind-protect functions, and I think we should simply take possession of the buffer at the start of the exception handler and put it back at the end, without copying anything. Whoever raises in between will reallocate a buffer. I see several implementation choices but I lack the data points that motivated the current implementation. So I will not explore this path right now. But if you can point me to further related discussions it would be useful to me.

In the meanwhile I propose to simply test on whether it goes on the minor or major heap; if it is the minor heap we already know that it is not going to raise because it fails with fatal_error. To me, the importance of having the no-raise guarantee justifies the deviation from normal style. If minor allocations are ever fixed to reliably raise, then it might be worthwhile to have _noexn variants of allocation functions in the public interface and then we can simplify this function.